src/validation.cpp

```diff
@@ -2000,11 +2027,70 @@ static int64_t nTimeTotal = 0;
 static int64_t nBlocksTotal = 0;
 
 /////////////////////////////////////////////////////////////////////// qtum
+bool CheckOpSender(const CTransaction& tx, const CChainParams& chainparams, int nHeight){
+    if(!tx.HasOpSender())
+        return true;
+
+    if(!(nHeight >= chainparams.GetConsensus().QIP5Height))
+        return false;
+
+    // Check that the sender address inside the output is only valid for contract outputs
+    for (const CTxOut& txout : tx.vout)
+    {
+        if(txout.scriptPubKey.HasOpSender() &&
+                !(txout.scriptPubKey.HasOpCreate() ||
+                  txout.scriptPubKey.HasOpCall()))
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
 bool CheckSenderScript(const CCoinsViewCache& view, const CTransaction& tx){
+    // Check for the sender that pays the coins
     CScript script = view.AccessCoin(tx.vin[0].prevout).out.scriptPubKey;
     if(!script.IsPayToPubkeyHash() && !script.IsPayToPubkey()){
```

Comment:
This rule could in theory be relaxed so that multisig or non-standard addresses can be used to pay for the gas of op_sender transactions. I think that there are other places this might break though if this rule is removed. (ie, gas refund code in the staker I think assumes pubkeyhash address). Needs discussion

Reply:
Yes. The sender used for the EVM and the sender used for refunds are different in case of `OP_SENDER` because the sender encoded with `OP_SENDER` does not provide coins. So with additional modifications can be provided other payers for the gas fee in case of `OP_SENDER`.

Reply:
I think it's not worth doing at this point, but should definitely support this in the next hardfork that this feature is updated in (probably x86)